src/TreeSitter/Importing.hs

```diff
@@ -1,135 +1,271 @@
+getFields = go Map.empty >>= \fields -> liftIO (print (Map.keys fields)) >> pure fields
+  where go fs = do
+          node <- peekNode
+          case node of
+            Just node' -> do
+              fieldName <- peekFieldName
+              keepGoing <- step
+              let fs' = case fieldName of
+                    Just fieldName' -> Map.insert fieldName' node' fs
+                    _ -> fs
+              if keepGoing then go fs'
+              else pure fs'
+            _ -> pure fs
 
--- | Return a 'ByteString' that contains a slice of the given 'Source'.
+-- | Return a 'ByteString' that contains a slice of the given 'ByteString'.
 slice :: Int -> Int -> ByteString -> ByteString
 slice start end = take . drop
   where drop = B.drop start
         take = B.take (end - start)
 
-class Importing type' where
 
-  import' :: (Alternative m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m type'
+newtype FieldName = FieldName { getFieldName :: String }
+  deriving (Eq, Ord, Show)
+
+
+-- | Generic construction of ASTs from a 'Map.Map' of named fields.
+--
+--   Product types (specifically, record types) are constructed by looking up the node for each corresponding field name in the map, moving the cursor to it, and then invoking 'buildNode' to construct the value for that field. Leaf types are constructed as a special case of product types.
+--
+--   Sum types are constructed by attempting to build each constructor nondeterministically. This should instead use the current node’s symbol to select the corresponding constructor deterministically.
+class GBuilding f where
+  gbuildNode :: (MonadFail m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m (f a)
+-- we'd only build the map when we know we're looking at a product
+
+
+instance GBuilding f => GBuilding (M1 D c f) where
+  gbuildNode = M1 <$> gbuildNode -- current node, not first child like above in original Building definition
+  -- gSymbolMatch _ = gSymbolMatch (Proxy @f)
+
+instance GBuilding f => GBuilding (M1 C c f) where
+  gbuildNode = M1 <$> gbuildNode
+
+-- Possibly for anonymous leaf nodes:
+instance GBuilding U1 where
+  gbuildNode = pure U1
+
+-- For regular leaf nodes
+instance {-# Overlappable #-} GBuilding (M1 S s (K1 c Text.Text)) where
+  gbuildNode = M1 . K1 <$> buildNode
+
+-- For unary products
+instance {-# Overlappable #-} (Selector s, Building k) => GBuilding (M1 S s (K1 c k)) where
+  gbuildNode = push $ do
+    fields <- getFields
+    gbuildProductNode fields
+
+-- For sum datatypes:
+instance (GBuildingSum f, GBuildingSum g, SymbolMatching f, SymbolMatching g) => GBuilding (f :+: g) where
+  gbuildNode = gbuildSumNode @(f :+: g)
+
+-- For product datatypes:
+instance (GBuildingProduct f, GBuildingProduct g) => GBuilding (f :*: g) where
+  gbuildNode = push $ do
+    fields <- getFields
```

Comment:
```suggestion
    getFields >>= gBuildProductNode @(f :+: g)
```

Reply:
Documenting this as a learning thing. When I tried to apply this suggestion with visible type application, I got this vague error message with some auto-generated `t1`:

```

src/TreeSitter/Importing.hs:227:37: error:
    • Cannot apply expression of type ‘t1’
      to a visible type argument ‘(f :+: g)’
    • In the second argument of ‘(>>=)’, namely
        ‘gBuildProductNode @(f :+: g)’
      In the second argument of ‘($)’, namely
        ‘getFields >>= gBuildProductNode @(f :+: g)’
      In the expression:
        push $ getFields >>= gBuildProductNode @(f :+: g)
    |
227 |   gbuildNode = push $ getFields >>= gBuildProductNode @(f :+: g)
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cabal: Failed to build tree-sitter-0.1.0.0 (which is required by
tree-sitter-python-0.1.0.1).
```

Removing the type application clarifies it, and it's about the identifier not being in scope: 
```
src/TreeSitter/Importing.hs:227:37: error:
    • Variable not in scope:
        gBuildProductNode :: Map.Map FieldName Node -> m ((:*:) f g a)
    • Perhaps you meant ‘gbuildProductNode’ (line 256)
    |
227 |   gbuildNode = push $ getFields >>= gBuildProductNode
    |                                     ^^^^^^^^^^^^^^^^^
cabal: Failed to build tree-sitter-0.1.0.0 (which is required by
tree-sitter-python-0.1.0.1).
```